iceoryx2-bb/posix/src/periodic_timer.rs

```diff
@@ -0,0 +1,235 @@
+//!
+//! let periodic_timer = PeriodicTimerBuilder::new().create().unwrap();
+//!
+//! // the timer waits on the following time points
+//! // 4 5 8 9 10 12 15 16 18
+//!
+//! periodic_timer.add(Duration::from_secs(4));
+//! periodic_timer.add(Duration::from_secs(5));
+//! periodic_timer.add(Duration::from_secs(9));
+//!
+//! std::thread::sleep(periodic_timer.next_iteration().unwrap());
+//!
+//! ```
+
+use std::{cell::RefCell, sync::atomic::Ordering, time::Duration};
+
+use iceoryx2_bb_log::fail;
+use iceoryx2_pal_concurrency_sync::iox_atomic::IoxAtomicU64;
+
+use crate::{
+    clock::ClockType,
+    clock::{Time, TimeError},
+};
+
+/// Represents an index to identify an added timer with [`PeriodicTimer::add()`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub struct PeriodicTimerIndex(u64);
+
+/// Represents the RAII guard of [`PeriodicTimer`] and is returned by [`PeriodicTimer::add()`].
+/// As soon as it goes out of scope it removes the attached cyclic timeout from [`PeriodicTimer`].
+pub struct PeriodicTimerGuard<'periodic_timer> {
+    periodic_timer: &'periodic_timer PeriodicTimer,
+    index: u64,
+}
+
+impl<'periodic_timer> PeriodicTimerGuard<'periodic_timer> {
+    /// Returns the underlying [`PeriodicTimerIndex`] of the attachment.
+    pub fn index(&self) -> PeriodicTimerIndex {
+        PeriodicTimerIndex(self.index)
+    }
+
+    /// Resets the attached timer and wait again the full time.
+    pub fn reset(&self) -> Result<(), TimeError> {
+        self.periodic_timer.reset(self.index)
+    }
+}
+
+impl<'periodic_timer> Drop for PeriodicTimerGuard<'periodic_timer> {
+    fn drop(&mut self) {
+        self.periodic_timer.remove(self.index);
+    }
+}
+
+/// Builder to create a [`PeriodicTimer`].
+pub struct PeriodicTimerBuilder {
+    clock_type: ClockType,
+}
+
+impl Default for PeriodicTimerBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl PeriodicTimerBuilder {
+    /// Creates a new builder.
+    pub fn new() -> Self {
+        Self {
+            clock_type: ClockType::default(),
+        }
+    }
+
+    /// Defines the [`ClockType`] that is used for time measurements. By default it is
+    /// [`ClockType::default()`].
+    pub fn clock_type(mut self, value: ClockType) -> Self {
+        self.clock_type = value;
+        self
+    }
+
+    /// Creates a new [`PeriodicTimer`]
+    pub fn create(self) -> Result<PeriodicTimer, TimeError> {
+        let start_time = fail!(from "PeriodicTimer::new()", when Time::now_with_clock(self.clock_type),
+                                "Failed to create PeriodicTimer since the current time could not be acquired.");
+        let start_time = start_time.as_duration().as_nanos();
+
+        Ok(PeriodicTimer {
+            timeouts: RefCell::new(vec![]),
+            id_count: IoxAtomicU64::new(0),
+            clock_type: self.clock_type,
+            last_iteration: RefCell::new(start_time),
+        })
+    }
+}
+
+#[derive(Debug)]
+struct Attachment {
```

Comment:
What's an `Attachment`? How about `Interval` or `ScheduledInterval`?

Reply:
It is internal API, therefore I wouldn't be too critical with the name.  It contains everything what the Waitset needs to store about something that was attached - `Attachment` sounded right.